compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/SIMTLayoutAnalysis.cpp

```diff
@@ -0,0 +1,266 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/LLVMGPU/Utils/SIMTLayoutAnalysis.h"
+#include "iree/compiler/Codegen/LLVMGPU/Utils/SIMTTransferFunctions.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Analysis/DataFlow/DeadCodeAnalysis.h"
+#include "mlir/Analysis/DataFlow/SparseAnalysis.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+#include "mlir/IR/AffineMap.h"
+
+#define DEBUG_TYPE "iree-simt-layout-analysis"
+
+using namespace mlir;
+using namespace mlir::iree_compiler;
+
+ChangeResult DistributionLayout::resolve(AffineMapLayout rhs) {
+  AffineMapLayout lhs = vectorLayout;
+
+  // If both layouts are same, do nothing.
+  if (lhs == rhs) {
+    return ChangeResult::NoChange;
+  }
+
+  // Take the more restrictive enforcement.
+  if (lhs.state < rhs.state) {
+    setState(rhs.state);
+    setLayout(rhs.layout);
+    return ChangeResult::Change;
+  } else if (lhs.state > rhs.state) {
+    return ChangeResult::NoChange;
+  }
+
+  // From here, both are in the same state, but have different layouts.
+
+  // WeaklyEnforced layouts don't need to be resolved.
+  if (lhs.state == Enforcement::WeaklyEnforced) {
+    return ChangeResult::NoChange;
+  }
+
+  // StronglyEnforced layouts need to be resolved.
+  // Layouts have a conflict. Insert a layout resolution operation.
+  llvm::errs() << "Layout conflict at: " << *this << "\n";
+  llvm::errs() << "With: " << rhs.layout << "\n";
+  assert(false && "Layout conflict");
+}
+
+ChangeResult DistributionLayout::resolve(const DistributionLayout *rhs) {
+  return resolve(rhs->vectorLayout);
+}
+
+void DistributionLayout::print(raw_ostream &os) const {
+  switch (getState()) {
+  case Enforcement::Uninitialized:
+    os << "Uninitialized";
+    break;
+  case Enforcement::WeaklyEnforced:
+    os << "WeaklyEnforced";
+    break;
+  case Enforcement::StronglyEnforced:
+    os << "StronglyEnforced";
```

Comment:
What is the distinction between strongly enforced and weakly enforced? For a vector.contract, I can imagine all of its operands would be strongly enforced, but would the surrounding element wise ops be strongly or weakly enforced?